### ISSUE TITLE
let_and_return : false positive

### ISSUE THREAD

**Issue by wagnerf42** [rust-lang/rust-clippy#1642]
false positive for let and return (possible linked to #848)
commented version works but generate clippy's warning.
given version does not compile:

```
error: `h` does not live long enough
 --> src/main.rs:9:1
  |
6 |     h.drain().map(|(_, v)| v).collect()
  |     - borrow occurs here
...
9 | }
  | ^ `h` dropped here while still borrowed
  |
  = note: values in a scope are dropped in the opposite order they are created
```

source code:

```rust
use std::collections::HashMap;
fn vecto() -> Vec<u32> {
    let mut h = HashMap::new();
    h.insert(1000, 3);
    h.insert(10000, 4);
    h.drain().map(|(_, v)| v).collect()
    //let r = h.drain().map(|(_, v)| v).collect();
    //r
}

fn main() {
    let v = vecto();
    println!("{:?}", v);
}
```

**Comment by montrivo** [comment]
The following block added in `let_return.rs` compiles and pass clippy tests. Some compiler issue must have been fixed.
```rust
use std::collections::HashMap;
fn vecto() -> Vec<u32> {
    #![allow(unreadable_literal)]
    let mut h = HashMap::new();
    h.insert(1000, 3);
    h.insert(10000, 4);
    h.drain().map(|(_, v)| v).collect()
    //let r = h.drain().map(|(_, v)| v).collect();
    //r
}

fn main() {
    let v = vecto();
    println!("{:?}", v);
}
```

**Comment by oli-obk** [comment]
@montrivo: I cannot confirm this. It is still broken on nightly: https://play.rust-lang.org/?gist=fa10281cc3811a555cda285c819db964&version=nightly

**Comment by montrivo** [comment]
Ok. Are the test files somehow not thoroughly compiled?

**Comment by oli-obk** [comment]
If an error happens in the AST, no HIR lints are reported

**Comment by ebroto** [comment]
This issue can be closed as the suggestion builds in current stable, see [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015)
